data/examples/declaration/class/associated-data-out.hs

```diff
@@ -1,8 +1,6 @@
 {-# LANGUAGE TypeFamilies #-}
 
-class Foo a where
-
```

Comment:
Is it possible to keep the old behavior here? I think putting `class` and `data` actually helps readability, or are there any concerns about idempotence or something else that makes it problematic?

Reply:
I also think it hurts readability to have single line data/class/instance declarations.

However, as you mentioned, there are idempotency concerns, ideally we should not convert any single line constructs to multi line ones. Here are some examples (hand-written, so  take them with a grain of salt):

```haskell
[d|class Foo a where data Foobar a|]
-- 1st pass:
[d|class Foo a where
  data Foobar a|]
-- 2nd pass:
[d| class Foo a where
  data Foobar a |]
```

```haskell
module M where { class Foo a where data Foobar a; foo = bar }
-- 1st pass
module M where { class Foo a where
  data Foobar a; foo = bar }
-- 2nd pass:
module M where

class Foo a  where
  data Foobar a

foo = bar
```

In short, since we change how we format things based on the fact that the enclosing SrcSpan is multi-line or not, we can not easily convert things to multi-line with the current design.

If we want to be able to do this kind of things, we can work-around this issue with some tweaks. For example, we can tweak the `located` combinator; so if we spit a newline inside the inner computation, we discard its effects and run it again as a mult-line layout (this should bubble up to enclosing `Located`'s as well). Actually, a similar fix can help us with unimplemented constructs as well; if we catch a "not implemented" error in a `Located` object, we can just use the original `SrcSpan` without formatting, and hope that it's still a valid syntax.

Reply:
Just to clarify; the reason we're formatting this in a single line is that the input is single line. We would format it as before if the input were multi-line.